sdks/java/io/jdbc/src/main/java/org/apache/beam/sdk/io/jdbc/JdbcIO.java

```diff
@@ -1634,12 +1637,22 @@ public void setup() throws Exception {
     }
 
     private Connection getConnection() throws SQLException {
-      Connection connection = this.connection;
-      if (connection == null) {
-        DataSource validSource = checkStateNotNull(this.dataSource);
-        connection = checkStateNotNull(validSource).getConnection();
-        this.connection = connection;
+      Connection connection;
+      DataSource validSource = checkStateNotNull(this.dataSource);
+      boolean reportLineage = false;
+      connectionLock.lock();
+      try {
+        connection = this.connection;
```

Comment:
I understand the race happened when creating connections. Is it necessary to put "connection = this.connection;" inside lock? This causes every getConnection() go through a lock (though tested it has little performance effect)

Or is it preferrable to only synchronize "connection = validSource.getConnection();" ?

Reply:
Done - thanks!